File: spring-boot-oauth2-auth-server/src/main/java/com/bittokazi/oauth2/auth/server/config/interceptors/TenantIdentifierInterceptorAdapter.java
package com.bittokazi.oauth2.auth.server.config.interceptors;

import com.bittokazi.oauth2.auth.server.app.repositories.master.TenantRepository;
import com.bittokazi.oauth2.auth.server.config.TenantContext;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.oauth2.jwt.JwtDecoder;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;

import java.util.Objects;

/**
 *
 * @author Bitto Kazi
 *
 */

@Component
public class TenantIdentifierInterceptorAdapter implements HandlerInterceptor {

    private final Logger logger = LoggerFactory.getLogger(this.getClass());
    private TenantRepository tenantRepository;

    private JwtDecoder jwtDecoder;

    public TenantIdentifierInterceptorAdapter(TenantRepository tenantRepository, JwtDecoder JwtDecoder) {
        this.tenantRepository = tenantRepository;
        this.jwtDecoder = jwtDecoder;
    }

    @Override
    public boolean preHandle(HttpServletRequest req, HttpServletResponse res, Object handler) throws Exception {
        String token = req.getHeader("Authorization").replace("Bearer ", "");
        return Objects.equals(jwtDecoder.decode(token).getClaims().get("tenant"), TenantContext.getCurrentTenant());
    }
}

File: spring-boot-oauth2-auth-server/src/main/kotlin/com/bittokazi/oauth2/auth/server/config/interceptors/TenantIdentifierInterceptorAdapter.kt
package com.bittokazi.oauth2.auth.server.config.interceptors

import com.bittokazi.oauth2.auth.server.app.repositories.master.TenantRepository
import com.bittokazi.oauth2.auth.server.config.TenantContext.getCurrentTenant
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.security.oauth2.jwt.JwtDecoder
import org.springframework.stereotype.Component
import org.springframework.web.servlet.HandlerInterceptor

/**
 *
 * @author Bitto Kazi
 */
@Component
class TenantIdentifierInterceptorAdapter(
    private val tenantRepository: TenantRepository,
    private val jwtDecoder: JwtDecoder
) : HandlerInterceptor {

    private val logger: Logger = LoggerFactory.getLogger(this.javaClass)

    @Throws(Exception::class)
    override fun preHandle(req: HttpServletRequest, res: HttpServletResponse, handler: Any): Boolean {
        val token = req
            .getHeader("Authorization")
            .replace(
                "Bearer ",
                ""
            )
        return jwtDecoder.decode(token).claims["tenant"] == getCurrentTenant()
    }
}

